### DISCUSSION TITLE
Knowledge distillation training performance

### DISCUSSION THREAD

**Discussion by abedkhooli** [lightonai/pylate#58] [Q&A] [answered]
Pylate looks easy and promising - great tool. 
I am trying to train a model using the [KD example](https://lightonai.github.io/pylate/documentation/training/) and the same [train data](lightonai/ms-marco-en-bge) of MS MARCO but Arabic (translated) version of queries and documents.  
Initial loss (losses.Distillation) starts in the range 0.05x and ends in 0.04x after one epoch - is this typical?   
I am using 250k queries with their corresponding scores. Base model: aubmindlab/bert-base-arabertv02

**Comment by NohTow** [comment, marked as answer]
Hello,
Thanks for your kind words!

Indeed, the loss is quite small ; this is because [the scores are normalized](https://github.com/lightonai/pylate/blob/d526d98b1b44899c2ba7e2c20addb22be2faecbd/pylate/losses/distillation.py#L112) (and so should be the teachers scores), [following the results of JaColBERTv2.5.](https://arxiv.org/pdf/2407.20750).
![image](https://github.com/user-attachments/assets/d313f16c-76a9-4223-a168-6a4d89520848)

This make me think that the default should be no normalization and only let as a (documented) option as it might be confusing (and sub-optimal) for people unaware.
I opened an issue to set the change the [default and will take care of this in the next days.](https://github.com/lightonai/pylate/issues/59) In the mean time, you can either also normalize the scores of the teacher in your dataset (ours is already normalized) or just set `normalize_scores` to `False` when creating the loss.

**Reply by abedkhooli** [reply]
This is very helpful. Thank you. 
Mine is not as smooth (green run still going, push_to_hub failed due to 2 process run - both creating same repo)
![image](https://github.com/user-attachments/assets/3d01fa1f-eba7-4497-a041-8a58dacd2427) 
I am actually using the same scores (lightonai/ms-marco-en-bge) which should be OK even if not fully applicable.

**Reply by NohTow** [reply]
In my own experiments, using the examples mined in English for other languages is ok, but you should score the query/document pairs in the language itself, which mean that you can use the dataset but might want to rescore it with an Arabic cross-encoder.
Yet, even though the results should be worse, the loss should look ok.
Maybe your LR is not adapted to your base model? If it's the same architecture as Bert-base, something along 2e-5 should be ok.
![image](https://github.com/user-attachments/assets/2e3c1fd3-e355-4e88-811b-2ba325dff490)

Edit: also, please note that you only have 80 steps, which might make it look more noisy just because my curves are smoothed

**Reply by abedkhooli** [reply]
I am actually using LR = 2e-5 and the base model is bert (64k vocab, not 32). Running for 1 epoch on 2T4s (Kaggle). Any recommendation for the number of epochs for this dataset (MS MARCO, bert base, 32 scores per query)?

**Reply by NohTow** [reply]
From my experiments, even though the loss is still going down after more than 1 epoch, the results are not improving (and actually degrading). It's a bit hard to get the scheduling right as the loss is not totally correlated with downstream results.
My suggestion would be to run actual downstream evals as often as you can and start by experimenting with only one epoch.

**Reply by abedkhooli** [reply]
Makes sense. Thanks.

**Reply by abedkhooli** [reply]
I completed training for 1 epoch but the final model (akhooli/ar_colbert_250_kd) seems indifferent against test queries/docs. A model trained via original ColBERT works fine on same test sample sentences (using Pylate retrieval). What did I miss?

**Reply by NohTow** [reply]
This is pretty though to say without much digging.
The model trained via Stanford is using our dataset aswell? This is most likely the issue, as once again, if you want to properly train the model, you need to rescore on the target language.

**Reply by abedkhooli** [reply]
Actually, the model I used to compare was trained with BM25 scores (available here: colbert-ir/colbertv2.0_msmarco_64way ). I thought the Pylate scores (https://huggingface.co/datasets/lightonai/ms-marco-en-bge/viewer/train) are more accurate and got encouraged with Pylate's ease of use.

**Reply by NohTow** [reply]
Indeed, BGE scores should be more accurate than BM25 (if both are computed on English and the domain is not too odd).
There seems to be something wrong in your training, but it's though to know what without more information.
Could you please check that the ids make sense w.r.t your data?

**Reply by abedkhooli** [reply]
I made a few random checks against queries and doc scores (query/doc/score) especially for extreme scores (1 or close, zero or close) and they all made sense (I filtered queries to remove those with non Arabic words and matched scores and docs). Do you have a trained KD model based on lightonai/ms-marco-en-bge I can check?

**Reply by NohTow** [reply]
We did not release any model for now.
The models I trained with the released dataset had results in-between [colbertv2.0](https://huggingface.co/colbert-ir/colbertv2.0) and [answerdotai/answerai-colbert-small-v1](https://huggingface.co/answerdotai/answerai-colbert-small-v1). ColBERTv2 was trained on the dataset you linked, so the dataset we released should be better.
If you want a model that works on Arabic data, there is [jinaai/jina-colbert-v2](https://huggingface.co/jinaai/jina-colbert-v2).

**Reply by abedkhooli** [reply]
I found that documents and queries in `lightonai/ms-marco-en-bge` are slightly different from those at `unicamp-dl/mmarco`. Also, not sure if ids being string vs int make a difference. After I aligned, filtered and cast ids it worked for a test case (1000 steps). 
There is an error in model-card creation while training and when saving the trained mode. It does not crash the process though but there is some time lost before skipping it. Not sure if this is sentence-transformers issue or Pylate params. 
`ERROR:pylate.models.colbert:Error while generating model card` 
`KeyError: 'scores'`
`File "/usr/local/lib/python3.10/dist-packages/pylate/models/colbert.py", line 1047, in _create_model_card`

**Reply by NohTow** [reply]
Yes indeed, you should use the documents and queries that are in `lightonai/ms-marco-en-bge`, they are in line with the ids with the ids related to scores. See this [example](https://lightonai.github.io/pylate/documentation/training/#knowledge-distillation-training).

> There is an error in model-card creation while training and when saving the trained mode. 

Yes, this is a known issue I need to fix, I procrastinated it as it is not crashing the training and just hurts the end "model card" on the hub, but I need to fix it at some point. This is sentence-transformers trying to create the model card but the code in PyLate remove some of the values needed for the card to make inference easier to read.

Happy to know that everything is working now!

**Reply by abedkhooli** [reply]
Thanks. For queries and documents I had to bring the corresponding Arabic versions. 
Do you think the model card error is from Pylate or Sentence Transformers?

**Reply by abedkhooli** [reply]
Are there recent changes to KD loss? 
I used a sentence transformer (SBERT) model to score docs against queries (MS Marco). When I train KD in Pylate the loss is almost constant at -3.46(07/8/9).

**Reply by NohTow** [reply]
Not really. 
I have been training a lot of ColBERT models with KD loss on the released datasets lately with a bunch of different base model arch and size and had no issue.
If you do not minmax-normalize the scores of your query/documents, you can try setting normalize_scores to False in the loss ([see #66](https://github.com/lightonai/pylate/pull/66)).

**Reply by abedkhooli** [reply]
Tried both T/F. Scores are cosine sims. The resulting model actually retrieves correctly with 100k samples and 1 epoch. Will score more and try.
